server/routes/companies.js

```diff
@@ -116,25 +179,399 @@ router.get('/filter', function(req, res, next) {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex.distinct('company_major.company_id').from('company_major')
+    .leftOuterJoin('company_position', 'company_major.company_id', 'company_position.company_id')
+    .where({position_id: req.body.pid, major_id: req.query.mid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else if (mid) {
-    db.query('SELECT company_id FROM company_major WHERE major_id = $1;', [mid], (err, result) => {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex('company_major').select('company_id').where({major_id: req.body.mid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else if (pid) {
-    db.query('SELECT company_id FROM company_position WHERE position_id = $1;', [pid], (err, result) => {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex('company_position').select('company_id').where({position_id: req.body.pid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else { 
-    return res.status(400).send({
-      message: '\'mid\' and \'pid\' are undefined'
-    });
+    util.throwError(400, '\'mid\' and \'pid\' are undefined');
   }
 });
 
+// Add a single company or related data
+router.post('/', function(req, res, next) {
+  values = {
+    name: req.body.name,
+    website: req.body.website,
+    logo: req.body.logo,
+    citizenship_requirement: req.body.citizenship_requirement,
+    description: req.body.description
+  }
+  let ranks = req.body.rank;
+  let user_ids = req.body.user_id;
+  let position_ids = req.body.position_id;
+  let major_ids = req.body.major_id;
+  let contact_ids = req.body.contact_id;
+  let event_ids = req.body.event_id;
+
+  console.log(req.body.position_id);
+
+  var queryCompany = knex('company').insert(values)
+  knex.transaction(function(trx) {
+    return queryCompany.transacting(trx)
+      .returning('id')
+      .then(async function(ids){
+        let company_id = ids[0];
+
+        if (ranks && user_ids) {
+          let rankValues = [];
+          if(Array.isArray(ranks) && Array.isArray(user_ids)){
+            let length = ranks.length;
+            if (ranks.length > user_ids.length)
+              length = user_ids.length; //not sure about error handling for length
+            for(let n=0; n < length; n++)
```

Comment:
Fix for loop styling:

for () {
  ...
}

instead of

for()
{
  ...
}

Reply:
done!